File: node/4-make-payment.ts
import dotenv from 'dotenv';
dotenv.config();
import { chromium } from "playwright-core";
import Browserbase from "@browserbasehq/sdk";
import { getCard } from './3-get-card';

const BROWSERBASE_API_KEY = process.env.BROWSERBASE_API_KEY!;
const BROWSERBASE_PROJECT_ID = process.env.BROWSERBASE_PROJECT_ID!;

const cardId = "ic_INPUT_CARD_ID_HERE"; // replace with your card id from the previous step

(async () => {
    const bb = new Browserbase({apiKey: BROWSERBASE_API_KEY});
    const session = await bb.sessions.create({
        projectId: BROWSERBASE_PROJECT_ID,
      });
    const browser = await chromium.connectOverCDP(session.connectUrl);
    const defaultContext = browser.contexts()[0];
    const page = defaultContext.pages()[0];

    const paymentInfo = await getCard(cardId);

    // log session ID url
    console.log(`Session URL: https://www.browserbase.com/sessions/${session.id}`);

    // go to the donation page
    await page?.goto("https://www.redcross.org/donate/donation.html");
    
    // click the first donation amount
    await page?.click("#modf-handle-0-radio");

    // click the continue button
    await page?.click("text=Continue");

    // click the credit card button
    await page?.click("text=credit card");

    // Fill billing information
    await page?.fill("input[name='bill_to_forename']", paymentInfo.cardholder_firstName!);
    await page?.fill("input[name='bill_to_surname']", paymentInfo.cardholder_lastName!);
    await page?.fill("input[name='bill_to_email']", paymentInfo.cardholder_email!);
    await page?.fill("input[name='bill_to_phone']", paymentInfo.cardholder_phone!);

    // Fill in the address information
    await page?.fill("input[name='bill_to_address_line1']", paymentInfo.cardholder_address.line1!);
    await page?.fill("input[name='bill_to_address_city']", paymentInfo.cardholder_address.city!);
    await page?.fill("input[name='bill_to_address_postal_code']", paymentInfo.cardholder_address.postal_code!);
    await page?.selectOption("select#bill_to_address_state", paymentInfo.cardholder_address.state!);

    // Fill in the card information
    await page?.fill("input#cardnumber", paymentInfo.card_number!);
    await page?.fill("input#MM", paymentInfo.expiration_month!.toString());
    await page?.fill("input#YY", paymentInfo.expiration_year!.toString());
    await page?.fill("input#CVC", paymentInfo.cvc!.toString());

    // click donate button
    await page?.click("text=Donate");

    await page.close();
    await browser.close();
})().catch((error) => console.error(error.message));

File: python/4-make-payment.py
from playwright.sync_api import Playwright, sync_playwright
from browserbase import Browserbase
from get_card import getCard
import os
import dotenv
dotenv.load_dotenv()

bb = Browserbase(api_key=os.environ["BROWSERBASE_API_KEY"])


cardId = "ic_INPUT_CARD_ID_HERE" # replace with your card id from the previous step

def run(playwright: Playwright) -> None:
    session = bb.sessions.create(
        project_id=os.environ["BROWSERBASE_PROJECT_ID"],
    )

    browser = playwright.chromium.connect_over_cdp(session.connect_url)
    context = browser.contexts[0]
    page = context.pages[0]

    payment_info = getCard(cardId)

    # Watch the session
    print(f"Session URL: https://browserbase.com/sessions/{session.id}")

    # Navigate to the donation page
    page.goto("https://www.redcross.org/donate/donation.html")

    # Perform actions on the donation page
    page.click("#modf-handle-0-radio")
    page.click("text=Continue")
    page.click("text=credit card")

    # Fill billing information
    page.fill("input[name='bill_to_forename']", payment_info["cardholder_firstName"])
    page.fill("input[name='bill_to_surname']", payment_info["cardholder_lastName"])
    page.fill("input[name='bill_to_email']", payment_info["cardholder_email"])
    page.fill("input[name='bill_to_phone']", payment_info["cardholder_phone"])

    # Fill in the address information
    page.fill("input[name='bill_to_address_line1']", payment_info["cardholder_address"]["line1"])
    page.fill("input[name='bill_to_address_city']", payment_info["cardholder_address"]["city"])
    page.fill("input[name='bill_to_address_postal_code']", payment_info["cardholder_address"]["postal_code"])
    page.select_option("select#bill_to_address_state", payment_info["cardholder_address"]["state"])

    # Fill in the card information
    page.fill("input#cardnumber", payment_info["card_number"])
    page.fill("input#MM", str(payment_info["expiration_month"]))
    page.fill("input#YY", str(payment_info["expiration_year"]))
    page.fill("input#CVC", str(payment_info["cvc"]))

    # Click donate button
    page.click("text=Donate")

    page.close()
    browser.close()

if __name__ == "__main__":
    with sync_playwright() as playwright:
        run(playwright)
